### PR TITLE
refactor: code review

### PR DESCRIPTION
**Pull request by hsk-kr**
Hi, I'm from the discord `Reactiflux` channel.

I have made some changes and they might not affect the production.
There was no problem with the test result, I recommend you to test the app yourself once more.

I will share my opinions based on my experiences, it would be thankful if you share your opinions with me too. 

- added .prettierrc to format files for readability
  In some files, two spaces as a tap was used, and in other files, four space as a tap was used.
  It would be better to use the same spaces. I used `prettier` to format files, but actually, it's up to you whether to use it or not.

- defined types for components' properties
  Actually, you can define the type for properties without defining them separately. I defined the types for each component's property type with name [Component Name]props.  

- fixed DisplayTire by changing params
  From what I have seen the path and the name, the function is supposed to be a component. I moved the event handler to props.

- added types RevsUnit, SidewallHeightUnit
  I thought it would be better to define the types in the type file to see all types in one file.

- Camelcased a type name from 'tire' -> 'Tire'

- used Omit to omit unused fields
  - DisplayTire
  - calculateTireHeight in helpers.tsx

  Since the `height` field was not used, it would be better to omit the field.

- used slice instead of conditional rendering
  When people who don't know the logic see the code, it would be more noticeable using the `slice` function. 

- removed expectable data types
  the following variables already have a type. So, I don't think that the types need to be defined. 

  ```typescript
    tires.map((tire, index: number) =>
  ```

  ```typescript
    const INCH_PER_MM = 0.03937008;
  ```

- used optional call instead of condition in TireHeightColumns
  You can use the optional operator to call if the function is callable.

- used if instead of using a ternary operator in a ternary operator in helpers
  using ternary operator is really useful and make the code shorter but in my opinion, using more than once can harm code readability. 

- changed prop name from 'handleX' -> 'onX'
  For consistency, it followed the prop name in `TireHeightColumn`.

Additionally, there are two things I want to mention.

First of all,
You can give hints for what code exactly does by giving a name.

For example,

**before**
```javascript
{(((tire1Height - tire2Height) / tire2Height) * speed + speed).toFixed(2)}
```

**after**
```javascript
{calculateX(tire1Height, tire2Height, speed).toFixed(2)}

// or you can use a variable

{X.toFixed(2)}
```

**before**
```javascript
 {range(10, 101, 10).map((speed) => {
  const tire1Height = calculateTireHeight(tires[0]);
  const tire2Height = calculateTireHeight(tires[1]);
  return (
    <td key={speed}>
    {(
       ((tire1Height - tire2Height) / tire2Height) * speed + speed ).toFixed(2)}
    </td>
    );
})}
```

**after**
```javascript
<SomethingRange tires={tires} speed={speed} />
```

**before**

```typescript
{tires[0] && tires[1] && (
  // ...
)}
```

**after**

```typescript
{hasBothTires && (
  // ...
)}
```

and last, you should use `useEffect` instead of `useMemo` in `App`.

`useMemo` is used for performance purposes, for example.

```javascript
const Comp = ({ value, list }) => {
  const convertedList = calculate(list); // it takes 1 sec every time

  return (
    <div>
      {value} / {convertedList.length}
    </div>
  );
}
```

Let's say, the calculate function takes 1 sec to return the result, and the `list` prop is barely changed while `value` is often changed.

Whenever the `value` is changed, the component will be rendered(called) and the `calculate` function will be called unnecessarily.

In this case, you can use `useMemo` to execute the logic only when `list` is changed.

```javascript
  const convertedList = useMemo(() => calculate(list), [list]);

  ...
```

---

This is it. These are my opinions.
Have a nice day, I will be waiting for your opinions.

### CODE REVIEW

File: src/components/DisplayTire.tsx [hunk 1]
```diff
@@ -1,11 +1,26 @@
-import {calculateTireHeight} from '../../util/helpers';
-import {tire} from "../../types/tire";
+import { calculateTireHeight } from '../../util/helpers';
+import { MouseEventHandler } from 'react';
+import { Tire } from '../../types/tire';
 
+interface DisplayTire extends Omit<Tire, 'height'> {
```
Comment (mkaatman) [inline src/components/DisplayTire.tsx:5]: Neat! I wasn't aware of Omit. https://www.typescriptlang.org/docs/handbook/utility-types.html#omittype-keys

File: src/components/DisplayTire.tsx [hunk 2]
```diff
@@ -1,11 +1,26 @@
...
-    return <span className="tire" onClick={() => handleClick()}>
-        <span>{`${width}/${aspectRatio}R`}</span>
-        <span style={{color: "#f00"}}>{wheelDiameter}</span>
-        <span>{` ⇔ ${calculateTireHeight({width, aspectRatio, wheelDiameter})}"`}</span>
-      </span>;
+export default function DisplayTire({
+  width,
+  aspectRatio,
+  wheelDiameter,
+  onClick,
```
Comment (mkaatman) [inline src/components/DisplayTire.tsx:13]: Much cleaner

File: .prettierrc [hunk 3]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (mkaatman) [inline .prettierrc:1]: Good idea for consistency.

File: src/components/SpeedometerComparisonTable.tsx [hunk 4]
```diff
@@ -1,26 +1,58 @@
-import {tire} from "../../types/tire";
-import {calculateTireHeight, range} from '../../util/helpers';
-import Card from "../ui/Card";
+import { Tire } from '../../types/tire';
+import { calculateTireHeight, range } from '../../util/helpers';
+import Card from '../ui/Card';
+
+interface SpeedometerComparisonTableProps {
```
Comment (mkaatman) [inline src/components/SpeedometerComparisonTable.tsx:5]: I like having an interface for the component props.

**Review by mkaatman — APPROVED** [review]
Thank you so much for the thorough review and ideas for improvements and maintainability. I really appreciate it! I had started with useEffect and switched but I think you're right it makes more sense than useMemo in this case.

**Comment by hsk-kr** [comment]
I'm glad it was helpful.
I have learned a lot from your code, thank you for the opportunity!
Have a nice coding!

**Comment by mkaatman** [comment]
VScode is giving me a warning on the way I'm importing App in my test. I did some searching but couldn't quite find a solution. It's not a big deal, the tests still run but I'm curious if you had any thoughts.

image

**Comment by mkaatman** [comment]
It looks like moving the tests folder inside of src fixes the issue. I'm not sure why.

**Comment by hsk-kr** [comment]
Thank you for sharing the issue! I learned a new one!
JSX needs `React` because actually JSX calls `React.createElement` to create an element.
By plugin react, you don't need to import `React` but it seems like it applies under the `src` folder by default.

**Comment by mkaatman** [comment]
I also did a refactor here if you're interested https://github.com/mkaatman/tirecomparecalc/pull/4/commits/cf8a23a90ca15459a0730e32f49966a89d6bf76d

**Pull request merged**